src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+
+   private:
+    void AddTyping(std::string arg_name, datalog::ArgumentType arg_type) {
+      if (!IsNameConstant(arg_name)) {
+        // This pass is only meant to find typings for constants,
+        // so do nothing if this argument is not a constant.
+        return;
+      }
+      if (type_env_.find(arg_name) == type_env_.end()) {
+        // No typing for this constant was found yet,
+        // so add it.
+        type_env_.insert({arg_name, arg_type});
+      } else if (type_env_.find(arg_name)->second != arg_type) {
+        // If there is already a typing for this constant,
+        // but it is different from the one used here,
+        // throw a type error and abort.
+        LOG(FATAL) << "type error for constant: " << arg_name;
+      }
+    }
+
+    // (TODO: #672) once an AST node for numbers as RVALUEs in numeric
+    // comparisons is added, this visitor should also visit
+    // the numeric RVALUEs and add them to the type environment
+    // with type ArgumentType(kNumber, "Number")
+    // This is a workaround that makes use of the fact that
+    // numeric comparisons are represented as predicates
+    // with a name that matches the operator:
+    // (https://github.com/google-research/raksha/blob/be6ef8e1e1a20735a06637c12db9ed0b87e3d2a2/src/ir/auth_logic/ast_construction.cc#L92)
+    static inline bool PredicateIsNumericOperator(
+        const datalog::Predicate& pred) {
+      if (pred.name() == "<" || pred.name() == "<" || pred.name() == "=" ||
+          pred.name() == "!=" || pred.name() == "<=" || pred.name() == ">=") {
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    Unit PreVisit(const Principal& principal) {
+      AddTyping(principal.name(), datalog::ArgumentType::MakePrincipalType());
+      return Unit();
+    }
+
+    Unit Visit(const datalog::Predicate& pred) {
+      if (PredicateIsNumericOperator(pred)) {
+        // This is part of the workaround for handling numeric comparisons
+        // given that the operators use the same AST nodes as predicates.
+        for (auto arg : pred.args()) {
+          AddTyping(arg, datalog::ArgumentType::MakeNumberType());
+        }
+        return Unit();
+      } else {
+        // This is the case where this is a normal predicate rather
+        // than a numeric operator
+        datalog::RelationDeclaration decl =
+            GetDeclarationOrFatal(decl_env_, pred.name());
+        // The relation declarations give the types for each position in the
+        // predicate. For the xth argument in the predicate, we want to
+        // assign it the type for the xth parameter in the relation declaration.
+        // So to do this, we iterate through the list of arguments in the
+        // predicate and the list of parameters in the declaration using
+        // the indices of the arguments iterator of the predicate.
+        for (unsigned i = 0; i < pred.args().size(); i++) {
```

Comment:
It's highly unlikely that this will ever be the case, but `unsigned` isn't long enough to cover all of the range of `pred.args().size()` on a 64 bit system, which is of type `size_t`. [`unsigned` is not preferred to be used anyway](https://google.github.io/styleguide/cppguide.html#Integer_Types). I'd say make `i` of type `size_t` or `uint64_t`.

Reply:
fixed